// containers/basket.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router';

import {
    removePhoneFromBasket,
    cleanBasket,
    basketCheckout
} from 'actions';
import { getPhonesWithCount, getTotalBasketPrices } from 'selectors';

class Basket extends Component {
    constructor(props) {
        super(props);
        this.renderContent = this.renderContent.bind(this);
        this.renderSidebar = this.renderSidebar.bind(this);
    }

    static propTypes = {
        phones: PropTypes.array.isRequired,
        totalPrice: PropTypes.number.isRequired,
        removePhoneFromBasket: PropTypes.func.isRequired,
        cleanBasket: PropTypes.func.isRequired,
        basketCheckout: PropTypes.func.isRequired
    }

    renderContent(phones, totalPrice, isBasketEmpty) {
        return (
            <div>
                {isBasketEmpty && <div>Корзина пуста</div>}

                <div className='table-responsive'>
                    <table className='table-bordered table-striped table-condensed cf'>
                        <tbody>
                            {phones.map((phone, index) => (
                                <tr
                                    key={index}
                                    className='item-checout'
                                >
                                    <td className='first-column-checkout'>
                                        <img
                                            className='img-thumbnail'
                                            src={phone.image}
                                            alt={phone.name}
                                        />
                                    </td>
                                    <td>{phone.name}</td>
                                    <td>${phone.price}</td>
                                    <td>{phone.count}</td>
                                    <td>
                                        <span
                                            className='delete-cart'
                                            onClick={() => this.props.removePhoneFromBasket(phone.id)}
                                        />
                                    </td>
                                </tr>
                            ))}
                        </tbody>
                    </table>
                </div>
                {
                    (!isBasketEmpty) &&
                    <div className='row'>
                        <div className='pull-right total-user-checkout'>
                            <b>Всего:</b>
                            ${totalPrice}
                        </div>
                    </div>
                }
            </div>
        );
    }

    renderSidebar(phones, isBasketEmpty) {
        return (
            <div>
                <Link
                    className='btn btn-info'
                    to='/'
                >
                    <span className='glyphicon glyphicon-info-sign' />
                    <span>Continue shopping!</span>
                </Link>
                {
                    (!isBasketEmpty) &&
                    <div>
                        <button
                            onClick={this.props.cleanBasket}
                            className='btn btn-danger'
                        >
                            <span className='glyphicon glyphicon-trash' />
                            Clear cart
                        </button>
                        <button
                            className='btn btn-success'
                            onClick={() => this.props.basketCheckout(phones)}
                        >
                            <span className='glyphicon glyphicon-envelope' />
                            Checkout
                        </button>
                    </div>
                }
            </div>

        );
    }

    render() {
        const { phones, totalPrice } = this.props;
        const isBasketEmpty = phones.length === 0;

        return (
            <div className='view-container'>
                <div className='container'>
                    <div className='row'>
                        <div className='col-md-9'>
                            {this.renderContent(phones, totalPrice, isBasketEmpty)}
                        </div>
                        <div className='col-md-3 btn-user-checkout'>
                            {this.renderSidebar(phones, isBasketEmpty)}
                        </div>
                    </div>
                </div>
            </div>

        )
    }
}

const mapStateToProps = state => ({
    phones: getPhonesWithCount(state),
    totalPrice: getTotalBasketPrices(state)
});

const mapDispatchToProps = {
    removePhoneFromBasket,
    cleanBasket,
    basketCheckout
};

export default connect(mapStateToProps, mapDispatchToProps)(Basket);
