// containers/login.js
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { browserHistory } from 'react-router';

import { loginAction } from 'actions';
import TextInput from 'component/TextInput';
import loginValidator from 'validator';

class Login extends Component {
    constructor(props){
        super(props);

        this.handleChange = this.handleChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleBlur = this.handleBlur.bind(this);

        this.state = {
            email: '',
            password: '',
            errors: {
                email: '',
                password: ''
            }
        };
    }

    static propTypes = {
        loginAction: PropTypes.func
    }

    handleChange(e) {
        this.setState({
            [e.target.name]: e.target.value
        });
    }

    handleBlur(e) {
        const error = loginValidator(e.target.name, e.target.value); 
        return this.setState({errors: Object.assign({}, this.state.errors, {
            [e.target.name]: error
        })});
    }

    async handleSubmit(e) {
        e.preventDefault();
        await this.props.loginAction(this.state);
        console.dir(this.props.location.state);
        if (!this.props.location.state) {
            browserHistory.push("/");
        } else {
            browserHistory.push(this.props.location.state.lastRoute);
        }
        
    }

    render() {
        return (
            <div className="col-md-4 col-md-offset-4 loginForm">
                <form onSubmit={this.handleSubmit}>
                    <TextInput 
                        type="email" 
                        name="email" 
                        label="Email"
                        placeholder="Enter email" 
                        onChange={this.handleChange}
                        onBlur={this.handleBlur}
                        errors={this.state.errors.email}
                    />
                    <TextInput 
                        type="password" 
                        name="password" 
                        label="Password"
                        placeholder="Password" 
                        onChange={this.handleChange}
                        onBlur={this.handleBlur}
                        errors={this.state.errors.password}
                    />
                    <button type="submit" className="btn btn-primary">Login</button>
                </form>
            </div>
        )
    }
}

const mapDispatchToProps = {
    loginAction
};

export default connect(null, mapDispatchToProps)(Login);
